### PR TITLE
add shape constraint ir design doc

### PR DESCRIPTION
**Pull request by wyzero**
to #325

### CODE REVIEW

**Review by qiuxiafei — DISMISSED** [review]
Nice documentation. Thanks~

File: docs/design/shape_constraint_ir.md [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+    ...
+  }];
+
+  let results = (outs HLO_Tensor:$result);
+}
+```
+
+## Injected by frontend converter
+
+BladeDISC Frontend converter lowers a coarse grained op to a bunch of fine grained MHLO ops. Not only the data computation but also the shape constraint should be lowered during dialect conversions.
```
[the remark this answers is not in the document]
Comment (wyzero) [reply docs/design/shape_constraint_ir.md]: done, thanks.

File: docs/design/shape_constraint_ir.md [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+  - `dim size % 4 = 0`;
+  - Likely values of the symbolic dimension (shape);
+  - Possible shape ranges of the symbolic dimension (shape);
+  - and many others ...
+
+# Where to get shape constraint?
+
+## Semantics of the definition of operation
+
+- The inputs and outputs of an element op of mhlo dialect should have same shape
```
[the remark this answers is not in the document]
Comment (wyzero) [reply docs/design/shape_constraint_ir.md]: done

File: docs/design/shape_constraint_ir.md [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+- Input shape of the model is dynamic
+  - different batch size in inference scenarios
+  - different image size, sequence length, etc
+- Shape is dynamic by design
+  - For seq2seq model, the number of decoding step is dynamic
+  - For CV detection model, the number of detected instants is dynamic
+  - For recommendation model, shape is affected by the value of its inputs (e.g. `tf.UniqueOp`).
+
+To solve the above problem, BladeDISC, an e2e dynamic shape compiler, is proposed. We already have seen many advantages of using BladeDISC over the previous static-shape-based compiler (e.g. XLA) in many pratical scenarios. However, there are still many rooms to improvement, especially in terms of the performance. In dynamic shape semantics, the concrete shape is out of reach, making many optimization strategies used in static shape semantics invalid. According to our previous experience in optimizing dynamic shape AI workloads, we found that shape constraint, the relationship between different symbolic shape (or dimension size), is very helpful in terms of performance optimization. For example:
```
[the remark this answers is not in the document]
Comment (wyzero) [reply docs/design/shape_constraint_ir.md]: done

**Review by qiuxiafei — APPROVED** [review]

**Pull request merged**